cpu/msp430_common/cpu.c

```diff
@@ -14,23 +14,46 @@
 #include "thread.h"
 
 /*
- * we must prevent the compiler to generate a prologue or an epilogue
- * for thread_yield_higher(), since we rely on the RETI instruction at the end
- * of its execution, in the inlined __restore_context() sub-function
+ * This function can both be called from ISR and from thread context.
+ *
+ * In both cases, the caller will use "CALL", which pushes the return
+ * address on the stack before executing the function's first instruction.
+ *
+ * If called within ISR, just set sched_context_switch_request and
+ * directly return to the call site. A regular "ret" does this.
+ *
+ * If called from stack context, it needs to prepare the stack so it looks exactly
+ * as if the thread had been interrupted by an ISR, which requires the SR to be on
+ * stack right after the return address. So we do this manually.
+ * __save_context() will then save the remaining registers, then store the stack
+ * pointer in the thread's thread_t.
+ *
+ * At this point, the thread context is properly saved. sched_run (possibly) changes
+ * the currently active thread, which __restore_context() then restores, resuming
+ * execution at the call site using reti.
+ *
+ * The naked attribute is needed to ensure this works.
  */
 __attribute__((naked)) void thread_yield_higher(void)
```

Comment:
Please drop the `__attribute__((naked))`, as

1. This is not applicable, the function is not solely consisting of bare assembly
2. The attribute needs to be given in the declaration of the function, not in the implementation.

Reply:
After some offline discussion regarding 2):

- the attribute does make a difference, on gcc 10.1.0 compiling for msp430, when the attribute is only given at the declaration.
In this case, the compiler added another "ret" instruction after the one that I had added explicitly, at the end of the if block.

I think in any case, as the function doesn't have any arguments (that are passed on stack), it wouldn't make any difference at the call site.

In the end, I removed the attribute.

Reply:
>     1. This is not applicable, the function is not solely consisting of bare assembly

That's not accurate, the "naked" attribute tells the compiler to not create function prologue or epilogue. It doesn't have to be purely basic asm.